### PR TITLE
Upgrade to Python 3.9 and add some of the newer Python magic we have

### PR DESCRIPTION
**Pull request by jon-betts**
For: https://github.com/hypothesis/client/issues/3375

 * This adds pinned versions of the docs dependencies, so they don't drift between local and CI
 * This is managed with `hdev requirements` so support for that has been added
 * Using the pyenv integration we have elsewhere so tox tracks the versions in the version file

## Review notes

This brings the Python portion of the client much more inline with our other projects and removes some possible cases where versions can drift.

## Test notes

 * With luck everything required here should automatically install itself upon running any `tox` command
 * If you want to try compiling the requirements you will need to install `hdev` as well - https://github.com/hypothesis/hdev
    * Then run `hdev requirements`
 * Other than that, everything should work as before:
    * `make docs`
    * `make checkdocs` - This fails with an indentation warning for me... but it also does in `master`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,11 +35,11 @@ lint: node_modules/.uptodate
 
 .PHONY: docs
 docs: python
-	tox -e py36-docs
+	tox -e docs
```
Comment (jon-betts) [inline Makefile:38]: We shouldn't need to specify the version here as we now use `tox-pyenv` will now find the version specified in `.python-version` when we don't specify.

File: tox.ini [hunk 2]
```diff
@@ -1,18 +1,19 @@
 [tox]
-envlist = py3-docs
+envlist = docs
 skipsdist = true
+requires =
+  tox-pip-sync
+  tox-pyenv
+  tox-run-command
```
Comment (jon-betts) [inline tox.ini:7]: In order this adds:

 * `tox-pip-sync` - Our new project to update your tox envs if the dependencies change
 * `tox-pyenv` - To make the version be determined by the `.python-version` file
 * `tox-run-command` - Required by `hdev` to work, but also quite handy as it lets you do `tox -e docs --run-command <ANYTHING YOU LIKE>`

File: tox.ini [hunk 3]
```diff
@@ -1,18 +1,19 @@
...
 
 [testenv]
 skip_install = true
 deps =
-    {docs,checkdocs}: recommonmark
-    {docs,checkdocs}: sphinx-autobuild
-    {docs,checkdocs}: sphinx
-    {docs,checkdocs}: sphinx_rtd_theme
+    {docs,checkdocs}: -r requirements/docs.txt
```
Comment (jon-betts) [inline tox.ini:12]: We now pin all our Python requirements for a few reasons:

 * So you can't get weirdness where it passes in CI but not locally or vice versa
 * So dependabot can poke us

File: tox.ini [hunk 4]
```diff
@@ -1,18 +1,19 @@
...
+    {env:EXTRA_DEPS:}
 passenv =
     HOME
-changedir =
-    {docs,checkdocs}: docs
```
Comment (jon-betts) [inline tox.ini:15]: Changing dirs isn't necessary, and makes `make requirements` write out local paths to the compiled files.

File: requirements/docs.txt [hunk 5]
```diff
@@ -0,0 +1,78 @@
+#
```
Comment (jon-betts) [inline requirements/docs.txt:1]: The file above is manually created, but this one is created by running `hdev requirements`

**Review by LMS007 — COMMENTED** [review]
Thanks Jon. Everything worked for me just fine

File: requirements/docs.txt (hunk 5 above)
Comment (robertknight) [reply requirements/docs.txt:1]: It would be great if the instructions at the top of this file could be updated to reflect the current process.

**Review by robertknight — APPROVED** [review]
LGTM. I can reproduce the issue with `make checkdocs` but as you said, that's unrelated to the changes on this branch.

File: requirements/docs.txt (hunk 5 above)
Comment (jon-betts) [reply requirements/docs.txt:1]: We actually can't do that unfortunately without creating problems with Dependabot. We both use `pip-compile` to create this file I think. So if we change the header Dependabot will change it back. This means a lot of PRs would have noise adding and removing a custom header.

**Pull request merged**